dusk-labyrinth.lic

```diff
@@ -7,74 +7,208 @@ custom_require.call %w[common common-items events]
       Flags.reset('pet')
-    elsif Flags['five-minute-warning'] || Flags['caught']
+    elsif Flags['five-minute-warning']
+      echo("*** Running out of time. Searching to find loot.") if $debug_mode_dl
+      search
+    elsif Flags['caught']
+      echo("*** We already caught a pet. Searching to find loot.") if $debug_mode_dl
       search
     end
+    # Put the goods away.
     stow_loot(loot_container)
+    # Move to another room.
     wander unless done?
   end
 
-  def next_move(exits)
-    if exits.count > 1 && !@prev_dir.nil?
-      exits.delete(@dir_to_prev_dir[@prev_dir.to_s])
+  # When the maze reshuffles then the room exits change.
+  # The XMLData.room_exits data is not refreshed so we
+  # literally don't know where we we can move to.
+  # Randomly move in directions until one works.
+  def wander_randomly
+    message("Maze reshuffled, trying to find a direction to move to")
+    ['n', 'ne', 'e', 'se', 's', 'sw', 'w', 'nw'].each do |dir|
+      if move(dir)
+        return
+      end
     end
-    exits.sample
+    message('Unable to move to another room after maze shuffled')
+    message('Try to move yourself to another room then restart script')
+    exit
   end
 
+  # Moves to a nearby room in the maze.
+  # Prioritizes rooms we have not been in yet.
+  # Skips over rooms we've been in before if looking for a pet.
   def wander
     waitrt?
-    next_dir = next_move(XMLData.room_exits)
-    @dir_to_prev_dir.each_key { |dir| break if move(dir) } unless move(next_dir)
-    @prev_dir = next_dir
+
+    @current_room = create_room if @current_room.nil?
+    echo("*** [wander] @current_room: #{@current_room}") if $debug_mode_dl
+
+    @forward_move = choose_next_move(@current_room['exits'])
+    @reverse_move = @reverse_direction_map[@forward_move]
+
+    echo("*** [wander] forward_move: #{@forward_move}") if $debug_mode_dl
+    echo("*** [wander] reverse_move: #{@reverse_move}") if $debug_mode_dl
+
+    move(@forward_move)
+
+    # Since rooms in the maze don't have ids, we build
+    # a bi-directional graph where each vertex (room)
+    # points to another vertex (room) and vice versa.
+    @previous_room = @current_room
+    @current_room = @current_room['exits'][@forward_move]['room'] || create_room
+
+    @previous_room['exits'][@forward_move]['room'] = @current_room
+    @current_room['exits'][@reverse_move]['room'] = @previous_room
+
+    # Increment the count of us having moved between these two rooms.
+    # In graph algorithms, this is known as a weighted graph so that
+    # when we pick the "next best move" to go down we choose paths
+    # with the least count (weight) thereby prioritizing paths we
+    # have not visited before spending time visiting rooms we've been to.
+    @previous_room['exits'][@forward_move]['count'] = @previous_room['exits'][@forward_move]['count'] + 1
+    @current_room['exits'][@reverse_move]['count'] = @current_room['exits'][@reverse_move]['count'] + 1
+
   end
 
+  def create_room
```

Comment:
In our in-memory graph of the maze, this function creates a new vertex (room) and initializes its internal state